### PR TITLE
Improve the propagation of color attributes over ConPTY

### LINKED ISSUE #2661: render: defer conversion of TextColor into COLORREF until actual rendering time (avoid ConPTY narrowing bugs)
Right now, every time Renderer needs to switch the brush colors, it does the following:

1. get the `TextColor` from the attribute run
1. get the RGB value of the color from the color table
1. pass that to `Engine::UpdateDrawingBrushes`

Step 2 is lossy, and is the source of a class of bugs I'll call "ConPTY narrowing" bugs. In addition, there's a step where `VtEngine` needs to _know the color table_ to do a translation from RGB back into indexed color.

**"narrowing" bugs**

|application output|conpty sends|why|
|-|-|-|
|`\e[38;2;22;198;12m`|`\e[92m`|RGB(22, 198, 12) matches color index 10 in Campbell|
|`\e[38;5;127m`|`\e[38;2;175;0;175m`|the xterm-256color color palette gets translated into RGB because Campbell only covers 0-16|
|`\e[38;2;12;12;12m`|`\e[30m`|RGB(12, 12, 12) matches index 0 in Campbell _and index 0 is the default color_ (#293)|

If Renderer passed the TextColor directly to the Engine, Xterm256Engine could pass through all indexed colors unchanged, and XtermEngine and WinTelnetEngine could flatten to the 16-color space.

* [ ] #3076 is caused by us doing early unpacking of colors and boldness; deferring until later will yield an improvement in preservation of underline status.

### LINKED ISSUE #3076: Underline not rendered after some SGR 22/39/49
# Environment

```none
Windows build number: Win32NT 10.0.18362.0
Windows Terminal version (if applicable): 0.5.2762.0
```

# Steps to reproduce

    echo -e '\e[4m underlined \e[33m yellow \e[39m should still be underlined \e[m'

# Expected behavior

The entire line should be underlined.

# Actual behavior

The "should still be underlined" bits aren't underlined.

Note that SGR codes `22`, `27`, `39` and `49` seem to cause this effect, but only if they're predeced by a `1`, `7`, `30-37` or `40-47`, resp., that is, only if they do actually turn off some other attribute.

Also note that if the "should still be underlined" text is made longer so that it overflows to the next line (or even multiple new lines), it's underlined in the last one of these lines. This makes it perhaps a reincarnation of #47.

Furthermore, changing the window width and thus causing a reflow of the text removes the underlining from the last line.

### LINKED ISSUE #3717: Unable to set colors above index 15 with OSC 4, take 2
# Environment

```none
Windows build number: 10.0.18362.0
Windows Terminal version (if applicable): 0.7.3291.0
```

# Steps to reproduce

Allegedly already fixed in #313, but doesn't work in v0.7:

    echo -ne '\e]4;15;rgb:ab/cd/ef\e\\'
    echo -ne '\e]4;16;rgb:ab/cd/ef\e\\'

# Expected behavior

Set palette index 15 and 16 to the given color.

# Actual behavior

Color index 15 is set to the given value (works for 0..15), index 16 remains unchanged (doesn't work for 16..255).

### LINKED ISSUE #5384: Should the bold/bright attribute apply to SGR 38/48 indexed colors?
# Environment

Windows build number: Version 10.0.18362.657
Windows Terminal version (if applicable): 0.10.781.0

# Steps to reproduce

1. Open a bash shell in conhost.
2. Execute `printf "\e[0;38;5;3mDARK \e[1mBRIGHT?\e[m\n"`
3. Open a bash shell in Window Terminal.
4. Execute `printf "\e[0;38;5;3mDARK \e[1mBRIGHT?\e[m\n"`

# Expected behavior

Both terminals should produce the same output showing the same colors.

# Actual behavior

Conhost displays the "BRIGHT?" text in the same dark shade as the "DARK" text. Windows Terminal displays it brighter.

image

In some ways this could be considered a narrowing bug (issue #2661), because one reason for the difference is that conhost converts the indexed color into an RGB value before storing it, while the Windows Terminal doesn't. And the bold/bright attribute has no effect on RGB colors, but does affect indexed colors.

That said, I don't think the WT implementation is correct, so fixing the "narrowing" issue in conhost is just going to make both of them wrong. There is some room for interpretation here, but I've been looking at how other terminals handle this, and AFAICT most (including XTerm and VTE) will only apply the brightness attribute to the 8 basic colors, and not the 256 indexed colors from SGR 38/48.

So this is really a question as much as it is a bug report. Which behavior do we want to follow here? Because the answer to that is going to affect how we address issue #2661.

### LINKED ISSUE #5864: New 0.11 VT attributes bug, attributes are not applied to " " (space) cells
In Windows Terminal 0.11.1251.0, a new VT attributes bug has been introduced.

When output contains VT escape sequences to set text attributes, they are sometimes ignored when the cells contains the space character.

If this sounds a lot like #5502, it is because it's probably exactly the same issue, you need to write spaces not only when the background color is different, but also if some other attributes, such underline are set (and you probably should make sure you include reverse video and crossed-out as well).

Repro command line : `curl 'https://raw.githubusercontent.com/PhMajerus/ANSI-art/master/Windows%20Terminal.ans' | sed 'H;$!d;x;s/\x1A.*$//' | iconv -f CP437`

image

Compare the top of the "W" with conhost:
image

It seems to be a bit random, even when it's not scrolling the buffer:
image

File used: https://github.com/PhMajerus/ANSI-art/blob/master/Windows%20Terminal.ans

### PR DESCRIPTION
**Pull request by j4james**
## Summary of the Pull Request

This PR reimplements the VT rendering engines to do a better job of preserving the original color types when propagating attributes over ConPTY. For the 16-color renderers it provides better support for default colors and improves the efficiency of the color narrowing conversions. It also fixes problems with the ordering of character renditions that could result in attributes being dropped.

## References

* This is only a partial fix for #293, but I suspect the remaining cases are unfixable.

## PR Checklist
* [x] Closes #2661
* [x] Closes #3076
* [x] Closes #3717
* [x] Closes #5384
* [x] Closes #5864
* [x] CLA signed.
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. Issue number where discussion took place: #2661

## Detailed Description of the Pull Request / Additional comments

Originally the base renderer would calculate the RGB color values and legacy/extended attributes up front, passing that data on to the active engine's `UpdateDrawingBrushes` method. With this new implementation, the renderer now just passes through the original `TextAttribute` along with an `IRenderData` interface, and leaves it to the engines to extract the information they need.

The GDI and DirectX engines now have to lookup the RGB colors themselves (via simple `IRenderData` calls), but have no need for the other attributes. The VT engines extract the information that they need from the `TextAttribute`, instead of having to reverse engineer it from `COLORREF`s.

The process for the 256-color Xterm engine starts with a check for default colors. If both foreground and background are default, it outputs a SGR 0 reset, and clears the `_lastTextAttribute` completely to make sure any reset state is reapplied. With that out the way, the foreground and background are updated (if changed) in one of 4 ways. They can either be a default value (SGR 39 and 49), a 16-color index (using ANSI or AIX sequences), a 256-color index, or a 24-bit RGB value (both using SGR 38 and 48 sequences).

Then once the colors are accounted for, there is a separate step that handles the character rendition attributes (bold, italics, underline, etc.) This step must come _after_ the color sequences, in case a SGR reset is required, which would otherwise have cleared any character rendition attributes if it came last (which is what happened in the original implementation).

The process for the 16-color engines is a little different. The target client in this case (Windows telnet) is incapable of setting default colors individually, so we need to output an SGR 0 reset if _either_ color has changed to default. With that out the way, we use the `TextColor::GetLegacyIndex` method to obtain an approximate 16-color index for each color, and apply the bold attribute by brightening the foreground index (setting bit 8) if the color type permits that.

However, since Windows telnet only supports the 8 basic ANSI colors, the best we can do for bright colors is to output an SGR 1 attribute to get a bright foreground. There is nothing we can do about a bright background, so after that we just have to drop the high bit from the colors. If the resulting index values have changed from what they were before, we then output ANSI 8-color SGR sequences to update them.

As with the 256-color engine, there is also a final step to handle the character rendition attributes. But in this case, the only supported attributes are underline and reversed video.

Since the VT engines no longer depend on the active color table and default color values, there was quite a lot of code that could now be removed. This included the `IDefaultColorProvider` interface and implementations, the `Find(Nearest)TableIndex` functions, and also the associated HLS conversion and difference calculations.

## Validation Steps Performed

Other than simple API parameter changes, the majority of updates required in the unit tests were to correct assumptions about the way the colors should be rendered, which were the source of the narrowing bugs this PR was trying to fix. Like passing white on black to the `UpdateDrawingBrushes` API, and expecting it to output the default `SGR 0` sequence, or passing an RGB color and expecting an indexed SGR sequence.

In addition to that, I've added some VT renderer tests to make sure the rendition attributes (bold, underline, etc) are correctly retained when a default color update causes an `SGR 0` sequence to be generated (the source of bug #3076). And I've extended the VT renderer color tests (both 256-color and 16-color) to make sure we're covering all of the different color types (default, RGB, and both forms of indexed colors).

I've also tried to manually verify that all of the test cases in the linked bug reports (and their associated duplicates) are now fixed when this PR is applied.

### CODE REVIEW

**Comment by DHowett** [comment]
Ah Christmas came early this year

File: src/renderer/vt/Xterm256Engine.cpp [hunk 1]
```diff
@@ -9,102 +9,79 @@ using namespace Microsoft::Console::Render;
...
+    }
 
-    hr = updateFlagAndState(ExtendedAttributes::CrossedOut,
-                            &Xterm256Engine::_BeginCrossedOut,
-                            &Xterm256Engine::_EndCrossedOut);
-    RETURN_IF_FAILED(hr);
+    if (textAttributes.IsReverseVideo() != _lastTextAttributes.IsReverseVideo())
+    {
+        RETURN_IF_FAILED(_SetReverseVideo(textAttributes.IsReverseVideo()));
+        _lastTextAttributes.SetReverseVideo(textAttributes.IsReverseVideo());
```
Comment (DHowett) [inline src/renderer/vt/Xterm256Engine.cpp:83]: This is one of those rare cases where I might _prefer_ a macro. This opinion is not likely held by the team (but I bet it could be socialized). What do you think? It's mechanical substring replacement, admittedly less debuggable, but repetition breeds potential for errors :smile:

```c++
#define UPDATE_STATE(stateName) \
 if(textAttributes.Is##stateName##() != _lastTextAttributes.Is##stateName##()) { \
  R_I_F(_Set##stateName##(...

UPDATE_STATE(CrossedOut)
UPDATE_STATE(Invisible)

#undef ...
```

**Comment by j4james** [comment]
> We'll need to update the WDDM and BGFX render engines in src/interactivity/onecore. 

OK, I'll try and update those. It looks like it should be a simple change, since both just seem to want the legacy attributes. Is there an easy way for me to build that code, at least enough to confirm that my changes are compiling?

> _Trivia: the WDDM renderer was copy/pasted to become the Dx renderer! I wonder if there's a world where we re-unify them. 

Well what's there now looks nothing like the Dx renderer, so if you want re-unify them, that would require somebody that knows what they're doing, which is definitely not me. 😀

**Comment by DHowett** [comment]
> Is there an easy way for me to build that code, at least enough to confirm that my changes are compiling?

Unfortunately, no. I'll handle any breaks that end up on the Windows side -- so don't gate too hard on it. Visual verification is enough for those two until we try to re-ingest the changes.

> definitely not me

Oh man, I wouldn't ask you to do that. Especially given that that project doesn't build in public :smile:

We've considered shipping some shim headers that would make those renderers compile -- ONLY compile -- just for basic validation.

File: src/renderer/vt/Xterm256Engine.cpp (hunk 1 above)
Comment (j4james) [reply src/renderer/vt/Xterm256Engine.cpp:83]: Personally I'm not a fan of macros, since I find it makes the code harder to understand. Like there's no indication what properties are being referenced or updated in that `UPDATE_STATE` call. But if you guys prefer that sort of thing, I don't mind changing it - I can see the argument for minimizing the repetition.

However, I should add that in the longer term I would have proposed rewriting much of that code anyway, to bundle the SGR sequences (e.g. instead of sending `\e[m\e[37m\e[1m\e[7m`, we should really be sending something like `\e[0;37;1;7m`). And in that case, the code would look completely different, and could well be one line per attribute anyway.

Comment (DHowett) [reply src/renderer/vt/Xterm256Engine.cpp:83]: I'm not terribly bothered by it. If in the long term we'll evolve towards attribute bundling, I'm down for leaving this expanded to make it painful/obvious.

**Comment by j4james** [comment]
FYI, I think this PR is now ready for review. But as we've discussed elsewhere, I think you'll probably want to hold off on merging until we have a fix for #5952 and the PowerShell problems.

I should also mention that this will result in WT losing support for the [`DECSCNM`](https://vt100.net/docs/vt510-rm/DECSCNM.htm) sequence, since the original implementation was conhost-only, and relied on the VT renderer passing through the translated colors, which doesn't happen anymore. I'll file a followup issue to deal with that.

**Comment by srisuk2519** [comment]
เมื่อ อา. 21 มิ.ย. 2563 เวลา 20:47 James Holderness <
notifications@github.com> เขียนว่า:

File: src/renderer/vt/Xterm256Engine.cpp (hunk 1 above)
Comment (miniksa) [reply src/renderer/vt/Xterm256Engine.cpp:83]: I vote no macro and for James's longer term plan of sending one SGR with semicolons that makes this mostly obsolete.

**Comment by miniksa** [comment]
> > _Trivia: the WDDM renderer was copy/pasted to become the Dx renderer! I wonder if there's a world where we re-unify them.
> 
> Well what's there now looks nothing like the Dx renderer, so if you want re-unify them, that would require somebody that knows what they're doing, which is definitely not me. 😀

> > definitely not me
> 
> Oh man, I wouldn't ask you to do that. Especially given that that project doesn't build in public 😄
> 
> We've considered shipping some shim headers that would make those renderers compile -- ONLY compile -- just for basic validation.

Yeah it'd be me. But I can't be bothered to do it at least until we get to go back into the office because setting up arcane, rarely used flavors of Windows is a pain in the butt and I'm not doing it without the ability to shout expletives inside my closed door office at work instead of at my home with my wife and son.

**Comment by miniksa** [comment]
Reading through it... 15/35...

Man, not passing through that interface and the color table is just SO MUCH CLEANER. I am loving it so far.

File: src/renderer/vt/paint.cpp [hunk 2]
```diff
@@ -185,144 +185,128 @@ using namespace Microsoft::Console::Types;
...
+    const auto needBold = fgIndex > 7;
+    if (needBold != _lastTextAttributes.IsBold())
+    {
+        RETURN_IF_FAILED(_SetBold(needBold));
+        _lastTextAttributes.SetBold(needBold);
+    }
 
-            _LastBG = colorBackground;
-        }
+    // After which we drop the hight bits, since only colors 0 to 7 are supported.
```
Comment (miniksa) [inline src/renderer/vt/paint.cpp:296]:
```suggestion
    // After which we drop the high bits, since only colors 0 to 7 are supported.

```

**Comment by DHowett** [comment]
I'm switching my signoff to a sign but adding a rejected build status; that'll more correctly represent what's going on here. :smile:

**Comment by j4james** [comment]
FYI, I've got a potential PR for #5952 which I'll probably commit tonight, but I won't be able to write up the PR details until tomorrow evening or the weekend. I'm not sure you'll like it, but it's at least an option to consider.
